### ISSUE TITLE
Alpine 1.21.4-alpine-ngx_http_datadog_module is not working with OpenResty 1.21.4 version

### ISSUE THREAD

**Issue by satishnaidu** [DataDog/nginx-datadog#4]
Hi Team,

Issue: Nginx/OpenResty container terminating after including load_module modules/ngx_http_datadog_module.so; without showing any error logs.

I would like to use Nginx with the Datadog module instead of using Nginx with the OpenTracing module.
I followed the steps in [Tracing the Proxy ](https://docs.datadoghq.com/tracing/trace_collection/proxy_setup/?_gl=1*16by6z6*_ga*MTM3MTM3Mjk0OS4xNjY2NzYyNzQ4*_ga_KN80RDFSQK*MTY2Njc2Mjc0OC4xLjEuMTY2Njc2Mjc2MC40OC4wLjA.&_ga=2.138428923.91409515.1666762748-1371372949.1666762748&tab=nginx#module-installation)

Nginx -V output:

`
/usr/local/openresty/nginx $ nginx -V
nginx version: openresty/1.21.4.1
built by gcc 10.3.1 20211027 (Alpine 10.3.1_git20211027) 
built with OpenSSL 1.1.1l  24 Aug 2021
TLS SNI support enabled
configure arguments: --prefix=/usr/local/openresty/nginx --with-cc-opt='-O2 -DNGX_LUA_ABORT_AT_PANIC -I/usr/local/openresty/pcre/include -I/usr/local/openresty/openssl/include' --add-module=../ngx_devel_kit-0.3.1 --add-module=../echo-nginx-module-0.62 --add-module=../xss-nginx-module-0.06 --add-module=../ngx_coolkit-0.2 --add-module=../set-misc-nginx-module-0.33 --add-module=../form-input-nginx-module-0.12 --add-module=../encrypted-session-nginx-module-0.09 --add-module=../srcache-nginx-module-0.32 --add-module=../ngx_lua-0.10.21 --add-module=../ngx_lua_upstream-0.07 --add-module=../headers-more-nginx-module-0.33 --add-module=../array-var-nginx-module-0.05 --add-module=../memc-nginx-module-0.19 --add-module=../redis2-nginx-module-0.15 --add-module=../redis-nginx-module-0.3.9 --add-module=../rds-json-nginx-module-0.15 --add-module=../rds-csv-nginx-module-0.09 --add-module=../ngx_stream_lua-0.0.11 --with-ld-opt='-Wl,-rpath,/usr/local/openresty/luajit/lib -L/usr/local/openresty/pcre/lib -L/usr/local/openresty/openssl/lib -Wl,-rpath,/usr/local/openresty/pcre/lib:/usr/local/openresty/openssl/lib' --with-pcre --with-compat --with-file-aio --with-http_addition_module --with-http_auth_request_module --with-http_dav_module --with-http_flv_module --with-http_geoip_module=dynamic --with-http_gunzip_module --with-http_gzip_static_module --with-http_image_filter_module=dynamic --with-http_mp4_module --with-http_random_index_module --with-http_realip_module --with-http_secure_link_module --with-http_slice_module --with-http_ssl_module --with-http_stub_status_module --with-http_sub_module --with-http_v2_module --with-http_xslt_module=dynamic --with-ipv6 --with-mail --with-mail_ssl_module --with-md5-asm --with-pcre-jit --with-sha1-asm --with-stream --with-stream_ssl_module --with-threads --with-stream --with-stream_ssl_preread_module

`

Please let me know how to proceed further on this.

**Comment by dgoffredo** [comment]
I'll need to put together a reproduction that allows me to get nginx into a debugger.

Will start looking into this next week at the latest.

**Comment by satishnaidu** [comment]
Thanks @dgoffredo let me know if you need any more information from my side.

**Comment by dgoffredo** [comment]
TL;DR I don't know what the underlying issue is yet.

Here's the status of my investigation:

- Created a feature branch [david.goffredo/openresty-crash][1] to play around with the module source in a new docker-compose setup.
[1]: https://github.com/DataDog/nginx-datadog/tree/david.goffredo/openresty-crash
- The crash happens here:
    ```
    Program received signal SIGSEGV, Segmentation fault.
    0x000072edf7a3fbb2 in ngx_set_env(opentracing::v3::string_view, ngx_cycle_s*) () from /usr/local/openresty/nginx/modules/ngx_http_datadog_module.so
    ```
    - this is a [function I wrote][2] to implicitly do the work of `env` directives to forward `DD_*` environment variables to worker processes.
- I don't know how to get the source files mapped into my container, so I did some printf debugging:
    ```
    (gdb) run
    Starting program: /usr/local/openresty/nginx/sbin/nginx -T
    warning: Error disabling address space randomization: Operation not permitted
    nginx: the configuration file /usr/local/openresty/nginx/conf/nginx.conf syntax is ok
    ()()()()() 275. entry: DD_AGENT_HOST
    ()()()()() 276. (void*)cycle: 0x72edf7bec1c0
    ()()()()() 277. (void*)ngx_get_conf(cycle->conf_ctx, ngx_core_module): 0x72edf7bee3a0
    ()()()()() 279. (void*)&ccf->env: 0x72edf7bee450
    ()()()()() 280. ccf->env.nelts: 126366385784522
    ()()()()() 281. ccf->env.size: 126366389298328
    ()()()()() 282. ccf->env.nalloc: 0
    ()()()()() 283. (void*)ccf->env.pool: 0x10
    ()()()()() 284. (void*)ccf->env.elts: 0x31
    ()()()()() 285. (void*)ccf->environment: 0x1
    ()()()()() 286. (void*)environ: 0x7fffcdb0bfe0
    ()()()()() 287. ccf->user: 0
    ()()()()() 288. ccf->user: 0
    ()()()()() 289. (void*)ccf->username: 0
    ()()()()() 300. after the ngx_array_push and null check. (void*)var: 0x98afd5f2ac79f021
    ()()()()() 303. after to_ngx_str(entry)

    Program received signal SIGSEGV, Segmentation fault.
    0x000072edf7a3fbb2 in ngx_set_env(opentracing::v3::string_view, ngx_cycle_s*) () from /usr/local/openresty/nginx/modules/ngx_http_datadog_module.so
    ```
- `ccf->env` contains bogus data (look at the stated length).
- The question is whether `ccf` is entirely bogus or if it's restricted to `ccf->env`. The trouble is that I don't know the contract for when which fields have their initial values.  For example, the `char** environment` is `0x1`, which makes me think that field is either unset or the whole `struct` is bogus.
- I double-checked that the docker image `openresty/openresty:1.21.4.1-3-alpine` does indeed use the same nginx sources as `
1.21.4-alpine-ngx_http_datadog_module.so.tgz`.
- `valgrind` further illustrates that the `ccf->env` `ngx_array` is bogus:
    ```
    ()()()()() 303. after to_ngx_str(entry)
    ==30== Invalid write of size 8
    ==30==    at 0x52EEBB2: ngx_set_env(opentracing::v3::string_view, ngx_cycle_s*) (in /usr/local/openresty/nginx/modules    /ngx_http_datadog_module.so)
    ==30==    by 0x52EED78: datadog_master_process_post_config(ngx_cycle_s*) (in /usr/local/openresty/nginx/modules/ngx_http_datadog_module.so)
    ==30==    by 0x160B9D: ngx_init_modules (ngx_module.c:72)
    ==30==    by 0x15DE45: ngx_init_cycle (ngx_cycle.c:635)
    ==30==    by 0x149DDB: main (nginx.c:295)
    ==30==  Address 0x17c86517982a61 is not stack'd, malloc'd or (recently) free'd
    ```

I made the following changes to the `nginx-datadog` source code to see if it changes the (crash) behavior:
- remove the module order in `module/config`
- remove all registered nginx configuration directives for `nginx-datadog`

Skipping the calls to `ngx_set_env` prevents the crash, and subsequent config processes succeeds, though I don't know if there is yet some sort of memory corruption.

I'm running out of ideas, but the investigation is ongoing.

One thing for me to try next is building `nginx-datadog` inside of the `openresty/openresty` image instead of in the corresponding `nginx/nginx` image.  I doubt that will make a difference, but it's worth a shot.

[1]: https://github.com/DataDog/nginx-datadog/tree/david.goffredo/openresty-crash
[2]: https://github.com/DataDog/nginx-datadog/blob/6e25f1b365b3bf745396423f037ddf10555b98fd/src/ngx_http_datadog_module.cpp#L250-L264

**Comment by dgoffredo** [comment]
The issue is that OpenResty's build of nginx contains patches that change the layout of a `struct` that our module is using.

I've found a way to avoid using that `struct`, and plan to have a fix tomorrow.

**Comment by dgoffredo** [comment]
Fixed in https://github.com/DataDog/nginx-datadog/pull/7, at least on Alpine.

**Comment by dgoffredo** [comment]
The latest release has the fix: https://github.com/DataDog/nginx-datadog/releases/tag/v0.3.5

**Comment by satishnaidu** [comment]
@dgoffredo  Thanks David for your time and fix.
